### ISSUE TITLE
unable to find an inherited method for function ‘imgRaster’ for signature ‘"VisiumV1"’

### ISSUE THREAD

**Issue by Prakrithi-P** [satijalab/seurat#5128]
Hi,
I am working with ST data from https://www.10xgenomics.com/resources/datasets/human-cerebellum-targeted-neuroscience-panel-1-standard-1-2-0 .

When I run SpatialFeaturePlot() , I get the following error:
Error in (function (classes, fdef, mtable)  : 
  unable to find an inherited method for function ‘imgRaster’ for signature ‘"VisiumV1"’

Also when I try out the data given in the vignette, 

InstallData("stxBrain")
trying URL 'http://seurat.nygenome.org/src/contrib/stxBrain.SeuratData_0.1.1.tar.gz'
Content type 'application/octet-stream' length 110346050 bytes (105.2 MB)

downloaded 1.5 MB

Error in download.file(url, destfile, method, mode = "wb", ...) : 
  download from 'http://seurat.nygenome.org/src/contrib/stxBrain.SeuratData_0.1.1.tar.gz' failed
In addition: There were 38 warnings (use warnings() to see them)
Warning in download.packages(pkgs, destdir = tmpd, available = available,  :
  download of package ‘stxBrain.SeuratData’ failed
Error in loadNamespace(name) : 
  there is no package called ‘stxBrain.SeuratData’

Help with any of the above issues would be great for me to start with ST data analysis.

Thanks,
Prakrithi

**Comment by allyhawkins** [comment]
I too am having this error. Uninstalling `SpatialExperiment` does the trick, but even when using the latest version on github, I still get this error when trying to create a `spatialFeaturePlot()` @lmweber. 

I have attached my session info for reference. 

<img width="277" alt="Screen Shot 2021-11-22 at 3 36 46 PM" src="https://user-images.githubusercontent.com/54039191/142938711-0b4beede-d912-4376-aa46-9e43de963c4b.png">

**Comment by lmweber** [comment]
Hi all, just following up since this issue came up again in discussions today.

I have tried to reproduce the error using the Seurat spatial vignettes, but those examples all seem to work fine for me.

Does anyone have an example / some code where this error occurs? This would be very helpful for tracking it down, and I would definitely like to help figure it out if this is still causing problems for people. (I had previously assumed it was related to older versions of Seurat / SpatialExperiment only, but it seems this may not be the case.)

**Comment by jaredslosberg** [comment]
This is still an issue for me, on R 4.1.2, Seurat 4.1.0, and SpatialExperiment 1.4.0.

If SpatialExperiment is loaded, I get the same error as everyone above. 

Following the vignette at https://satijalab.org/seurat/articles/spatial_vignette.html, I get this error:

```
> SpatialFeaturePlot(brain, features = "nCount_Spatial") + theme(legend.position = "right")
Error in (function (classes, fdef, mtable)  : 
  unable to find an inherited method for function ‘imgRaster’ for signature ‘"VisiumV1"’
In addition: Warning message:
`guides(<scale> = FALSE)` is deprecated. Please use `guides(<scale> = "none")` instead. 
```
And here is my session info (with SpatialExperiment loaded)
![image](https://user-images.githubusercontent.com/59885637/152430969-82838d8e-377f-4709-9d0e-01653abb120f.png)

And here is the traceback:
```
31: stop(gettextf(“unable to find an inherited method for function %s for signature %s”, 
    sQuote(fdef@generic), sQuote(cnames)), domain = NA)
30: (function (classes, fdef, mtable) 
  {
    methods <- .findInheritedMethods(classes, fdef, mtable)
    if (length(methods) == 1L) 
      return(methods[[1L]])
   ...
29: imgRaster(x)
28: dim(x)
27: dim(x)
26: ncol(x = image)
25: data.frame(x = c(0, ncol(x = image)), y = c(0, nrow(x = image)))
24: transform_position(data, panel_params$x$rescale, panel_params$y$rescale)
23: f(...)
22: coord$transform(data.frame(x = c(0, ncol(x = image)), y = c(0, 
    nrow(x = image))), panel_scales)
21: f(...)
20: (structure(function (...) 
  f(...), class = “ggproto_method”))(data, panel_params, coord, 
    image = new(“VisiumV1”, image = structure(c(0, 0, 0, 0, 0, 
    0, 0, 0, 0, 0.43921568627451, 0.901960784313726, 0.894117647058824, 
    0.905882352941176, 0.905882352941176, 0.909803921568627, 
   ...
19: do.call(self$draw_panel, args)
18: FUN(X[[i]], ...)
17: lapply(split(data, data$PANEL), function(data) {
    if (empty(data)) 
      return(zeroGrob())
    panel_params <- layout$panel_params[[data$PANEL[1]]]
    do.call(self$draw_panel, args)
   ...
16: f(..., self = self)
15: self$geom$draw_layer(data, self$computed_geom_params, layout, 
    layout$coord)
14: f(..., self = self)
13: l$draw_geom(d, layout)
12: (function (l, d) 
  l$draw_geom(d, layout))(dots[[1L]][[1L]], dots[[2L]][[1L]])
11: mapply(FUN = f, ..., SIMPLIFY = FALSE)
10: Map(function(l, d) l$draw_geom(d, layout), plot$layers, data)
9: ggplot_gtable.ggplot_built(ggplot_build(x))
8: ggplot_gtable(ggplot_build(x))
7: ggplotGrob(x)
6: plot_table.ggplot(X[[i]], ...)
5: FUN(X[[i]], ...)
4: lapply(x$plots, plot_table, guides = guides)
3: build_patchwork(plot, plot$layout$guides %||% “auto”)
2: print.patchwork(x)
1: (function (x, ...) 
  UseMethod(“print”))(x)
```

**Comment by lmweber** [comment]
Ok thanks. After looking through the Seurat and SpatialExperiment code some more now, I think this is related to the fact that both Seurat and SpatialExperiment have a class called `SpatialImage`. So if both packages are installed, Seurat is trying to use the wrong `SpatialImage`.

Not sure what the best solution is yet, but will think about it some more. Thanks for raising it.

**Comment by lmweber** [comment]
Yes, I think this is due to both packages having a class called `SpatialImage`, which is an unfortunate coincidence.

So R is probably choosing which class to use depending on which order the two packages are loaded with `library()`, which is also why this error is only happening sometimes and not other times.

We will discuss this some more to try to find a solution. Thanks all for your comments.

**Comment by HelenaLC** [comment]
For completeness, I could reproduce the error as follows:
```
suppressPackageStartupMessages({
    library(Seurat)
    library(SeuratData)
    library(SeuratObject)
    library(SpatialExperiment)
})

if (!"stxBrain" %in% InstalledData()$Dataset)
    InstallData("stxBrain")
data("anterior1")

SpatialFeaturePlot(
    object = anterior1, 
    features = rownames(anterior1)[1])

> Error in (function (classes, fdef, mtable)  : 
  unable to find an inherited method for function ‘imgRaster’ for signature ‘"VisiumV1"’
```
We will push a fix on the `SpatialExperiment`-end shortly.

**Comment by lmweber** [comment]
This has now been addressed in the latest update to the `SpatialExperiment` package. We renamed our `SpatialImage` class to `VirtualSpatialImage`, so now there is no longer a clash with the `SpatialImage` class name in Seurat.

The update is available in `SpatialExperiment` version 1.5.3, which can be installed from [GitHub](https://github.com/drighelli/SpatialExperiment). This version is now going through the Bioconductor build system, so should also be available via Bioc-devel from later this week onwards (Wed or Thu).

Additional details in the `SpatialExperiment` repo [here](https://github.com/drighelli/SpatialExperiment/issues/98) and [here](https://github.com/drighelli/SpatialExperiment/pull/101).

Thanks all for your feedback! Glad we were able to figure this out.
